### PR TITLE
strict-input-validation for playground without test cases

### PR DESCRIPTION
**Pull request by AnandAili**

### CODE REVIEW

File: src/main/java/smokefree/domain/Initiative.java [hunk 1]
```diff
@@ -50,6 +57,9 @@
 
     @CommandHandler
     public Initiative(CreateInitiativeCommand cmd, MetaData metaData) {
+        validateMaximumPlaygroundCapacity();
+        validateDuplicatePlaygroundNames(cmd.getName());
+        validatePlaygroundsRange(cmd.getGeoLocation(), SmokefreeConstants.MAXIMUM_PLAYGROUNDS_DISTANCE);
```
Comment (dimitriirtimid) [inline src/main/java/smokefree/domain/Initiative.java]: Note the inconsistency here. The MAXIMUM_PLAYGROUNDS_ALLOWED is accessed in the validation method, while the MAXIMUM_PLAYGROUNDS_DISTANCE is passed in as a parameter. 
Is there a reason for this?

File: src/main/java/smokefree/domain/Initiative.java [hunk 2]
```diff
@@ -206,5 +216,46 @@ private void assertCurrentUserIsManager(MetaData metaData) {
                 "You are not a manager of this playground"));
     }
 
+    private void validateMaximumPlaygroundCapacity() {
+        if(initiativeProjection.playgrounds().size() >= initiativeProjection.MAXIMUM_PLAYGROUNDS_ALLOWED) {
+            throw new DomainException("MAXIMUM_PLAYGROUNDS_CAPACITY_REACHED",
+                    "Can not add more than " + initiativeProjection.MAXIMUM_PLAYGROUNDS_ALLOWED + " playgrounds",
```
Comment (dimitriirtimid) [inline src/main/java/smokefree/domain/Initiative.java]: initiativeProjection.MAXIMUM_PLAYGROUNDS_ALLOWED don't put these constants in the projection. They should not have any validation related functionality.

File: src/main/java/smokefree/domain/Initiative.java [hunk 3]
```diff
@@ -206,5 +216,46 @@ private void assertCurrentUserIsManager(MetaData metaData) {
...
+
+    private void validateDuplicatePlaygroundNames(String playgroundName) {
+        initiativeProjection.playgrounds().entrySet()
+                .stream()
+                .filter( playgroundEntry -> playgroundEntry.getValue().getName().equals(playgroundName))
+                .map(playgroundEntry -> playgroundEntry.getValue())
+                .findFirst()
+                .ifPresent( p -> {
+                    throw new DomainException("DUPLICATE_PLAYGROUND_NAME",
+                            "Playground " + playgroundName + " does already exist, please choose a different name",
```
Comment (dimitriirtimid) [inline src/main/java/smokefree/domain/Initiative.java]: I am getting the feeling you are mixing up the order of the parameters in the DomainException. The second parameter is the technical message that is not sent to the front-end and the third is a message that is sent to the front-end. But your second message is more user-friendly than the third. You perhaps you intended them to be the other way around?

File: src/main/java/smokefree/domain/Initiative.java [hunk 4]
```diff
@@ -206,5 +216,46 @@ private void assertCurrentUserIsManager(MetaData metaData) {
...
+                .filter( playgroundEntry -> {
+                    Playground playground = playgroundEntry.getValue();
+                    GlobalPosition currentPlaygroundPosition = new GlobalPosition(playground.getLat() , playground.getLng(), 0.0);
+                    double playgroundsDistance = geodeticCalculator.calculateGeodeticCurve(ellipsoidsReference, currentPlaygroundPosition, newPlaygroundPosition).getEllipsoidalDistance();
+                    return  playgroundsDistance < distance ? true:false;
+                }).map(playgroundEntry -> playgroundEntry.getValue())
+                .findFirst()
+                .ifPresent(p -> {
+                    throw new DomainException("PLAYGROUNS_LOCATED_CLOSELY",
+                            "Two playgrounds can not exist within 100 Meters",
```
Comment (dimitriirtimid) [inline src/main/java/smokefree/domain/Initiative.java]: hmm 100 ? Something is wrong here. Please see for yourself what.

File: src/main/java/smokefree/domain/Initiative.java [hunk 5]
```diff
@@ -206,5 +216,46 @@ private void assertCurrentUserIsManager(MetaData metaData) {
...
+    private void validatePlaygroundsRange(GeoLocation newPlaygroundLocation, long distance) {
+        GeodeticCalculator geodeticCalculator = new GeodeticCalculator();
+        Ellipsoid ellipsoidsReference = Ellipsoid.WGS84;
+        GlobalPosition newPlaygroundPosition = new GlobalPosition(newPlaygroundLocation.getLat(), newPlaygroundLocation.getLng(), 0.0); // Point A
+        initiativeProjection.playgrounds().entrySet().stream()
+                .filter( playgroundEntry -> {
+                    Playground playground = playgroundEntry.getValue();
+                    GlobalPosition currentPlaygroundPosition = new GlobalPosition(playground.getLat() , playground.getLng(), 0.0);
+                    double playgroundsDistance = geodeticCalculator.calculateGeodeticCurve(ellipsoidsReference, currentPlaygroundPosition, newPlaygroundPosition).getEllipsoidalDistance();
+                    return  playgroundsDistance < distance ? true:false;
```
Comment (dimitriirtimid) [inline src/main/java/smokefree/domain/Initiative.java]: why `? true:false` it will work but is superfluous?

File: src/main/java/smokefree/graphql/CreateInitiativeInput.java [hunk 6]
```diff
@@ -4,18 +4,40 @@
...
+import javax.validation.constraints.NotNull;
+import javax.validation.constraints.Size;
+
 @Getter
 @NoArgsConstructor(access = AccessLevel.PRIVATE)
 @AllArgsConstructor
 public
 class CreateInitiativeInput {
+
+    @NotNull(message = "The initiativeId must have a value")
```
Comment (dimitriirtimid) [inline src/main/java/smokefree/graphql/CreateInitiativeInput.java]: Very nice, but I think actually a bit too elaborate. We have a formal interface, so these errors should never occur. So I don't think we need to specify detailed error message. Also some indication is probably given by the validation framework?

File: src/main/java/smokefree/graphql/CreateInitiativeInput.java [hunk 7]
```diff
@@ -4,18 +4,40 @@
...
+    @Size(min = 1, message = "The initiativeId must not be blank")
     String initiativeId;
+
     Type type;
+
+    @Enumerated(EnumType.STRING)
```
Comment (dimitriirtimid) [inline src/main/java/smokefree/graphql/CreateInitiativeInput.java]: Question: What is the purpose of this line?

File: src/main/java/smokefree/graphql/CreateInitiativeInput.java [hunk 8]
```diff
@@ -4,18 +4,40 @@
...
+    @NotNull(message = "The status must not be blank")
+    //@Pattern(regexp = "a|b|c|d", flags = Pattern.Flag.CASE_INSENSITIVE) // TO-DO: Need to test it for possible values during integration test
     Status status;
+
+    @Length(min = 3, message = "The name must be at least 3 characters")
```
Comment (dimitriirtimid) [inline src/main/java/smokefree/graphql/CreateInitiativeInput.java]: I like these messages

File: src/main/java/smokefree/graphql/GraphqlFactory.java [hunk 9]
```diff
@@ -51,9 +52,18 @@
 
     @Bean
     public ErrorExtensionsMapper<ValidationException> mapValidationExceptionToErrorExtension() {
-        return exceptionToErrorExtensionsMapper(ValidationException.class, e -> new ErrorExtensions(ErrorCode.VALIDATION, "Invalid data"));
+        return exceptionToErrorExtensionsMapper(ValidationException.class, e -> new ErrorExtensions(ErrorCode.VALIDATION, e.getMessage()));
```
Comment (dimitriirtimid) [inline src/main/java/smokefree/graphql/GraphqlFactory.java]: good job!

File: src/main/java/smokefree/graphql/GraphqlFactory.java [hunk 10]
```diff
@@ -51,9 +52,18 @@
...
+    }
+
+    @Bean
+    public ErrorExtensionsMapper<ExecutionException> mapDomainExceptionToErrorExtensions() {
+        return exceptionToErrorExtensionsMapper(ExecutionException.class, e -> new ErrorExtensions(ErrorCode.VALIDATION, e.getMessage().split(":")[1]));
```
Comment (dimitriirtimid) [inline src/main/java/smokefree/graphql/GraphqlFactory.java]: Let's discuss. I am curious where these exceptions come from and also whether we do not expose too much information if we send their message to the frontend

File: src/main/java/smokefree/projection/InitiativeProjection.java [hunk 11]
```diff
@@ -26,9 +20,30 @@
 @Slf4j
 @Singleton
 public class InitiativeProjection {
+
+    /**
+     * It holds a maximum value that System can allow user to add playgrounds
+     */
+    public final long MAXIMUM_PLAYGROUNDS_ALLOWED;
```
Comment (dimitriirtimid) [inline src/main/java/smokefree/projection/InitiativeProjection.java]: as stated: this is not the place

File: src/main/java/smokefree/projection/InitiativeProjection.java [hunk 12]
```diff
@@ -26,9 +20,30 @@
...
+    public InitiativeProjection() {
+        MAXIMUM_PLAYGROUNDS_ALLOWED = SmokefreeConstants.MAXIMUM_PLAYGROUNDS_ALLOWED;
+    }
+
+    /**
+     * the custom value for the maximum playgrounds to be allowed into the System
+     * @param maximumPlaygrounds maximum number of playgrounds allowed
+     */
+    public InitiativeProjection(Long maximumPlaygrounds) {
+        MAXIMUM_PLAYGROUNDS_ALLOWED = maximumPlaygrounds;
```
Comment (dimitriirtimid) [inline src/main/java/smokefree/projection/InitiativeProjection.java]: why make it this complicated?

File: src/main/java/smokefree/projection/InitiativeProjection.java [hunk 13]
```diff
@@ -145,4 +160,9 @@ public Progress progress() {
         return progress;
     }
 
+
+    public Map<String, Playground> playgrounds() {
```
Comment (dimitriirtimid) [inline src/main/java/smokefree/projection/InitiativeProjection.java]: I don't like this style. Can you make it something like getAllPlaygrounds() ? 
Also I would just return the values

File: src/test/java/smokefree/domain/CreateInitiativeCommandTest.java [hunk 14]
```diff
@@ -27,4 +28,110 @@ void when_html_input_should_fail() {
...
+                Status.not_started,
+                "Happy Diemen",
+                null,
+                32.234566
+        );
+        Set<ConstraintViolation<CreateInitiativeInput>> violations = validatorFactory.getValidator().validate(cmd);
+        assertEquals(1, violations.size());
+        assertEquals("The lat must not be blank", violations.iterator().next().getMessage());
+
+    }
```
Comment (dimitriirtimid) [inline src/test/java/smokefree/domain/CreateInitiativeCommandTest.java]: nice set of validations!

File: src/main/java/chatbox/ChatboxController.java [hunk 15]
```diff
@@ -31,22 +32,22 @@
...
 
     @Post("/{chatboxId}")
-    public String postMessage(Authentication authentication, String chatboxId, @Size(max=4096) @Body ChatMessage chatMessage) {
+    public ChatMessage postMessage(Authentication authentication, String chatboxId, @Size(max=4096) @Body ChatMessage chatMessage) {
         final String userName = authentication.getAttributes().get("cognito:username").toString();
         log.info("chat message for"  + chatboxId + ": " + chatMessage + " from: " + userName);
 
         if (!isValidChatboxId(chatboxId))
-            return "error: invalid chatbox";
+            throw new ValidationException("error: invalid chatbox");
```
[the remark this answers is not in the document]
Comment (AnandAili) [reply src/main/java/chatbox/ChatboxController.java]: Yes, due to 'pull --rebase', its appearing in this branch as well

File: src/main/java/smokefree/Mutation.java [hunk 16]
```diff
@@ -95,6 +98,10 @@ public InputAcceptedResponse decideToBecomeSmokeFree(DecideToBecomeSmokeFreeComm
     }
 
     public InputAcceptedResponse decideToNotBecomeSmokeFree(DecideToNotBecomeSmokeFreeCommand cmd, DataFetchingEnvironment env) {
+        // the status can not be stopped unless the playground managers choose that the playground can not be smokefree
```
[the remark this answers is not in the document]
Comment (AnandAili) [reply src/main/java/smokefree/Mutation.java]: this comment appeared here due to 'pull --rebase' from the master. And I will pick it up in the next set of validation as we discussed earlier.

File: src/main/java/smokefree/Mutation.java [hunk 17]
```diff
@@ -95,6 +98,10 @@ public InputAcceptedResponse decideToBecomeSmokeFree(DecideToBecomeSmokeFreeComm
...
+        if(initiativeProjection.playground(cmd.getInitiativeId(), toContext(env).requireUserId()).getStatus().equals(Status.stopped)) {
+            throw new DomainException("PLAYGROUND_INITIATIVE_ALREADY_STOPPED", "The Initiative for this playground is already stopped" , "Please contact help line for more details");
```
[the remark this answers is not in the document]
Comment (AnandAili) [reply src/main/java/smokefree/Mutation.java]: made changes as suggested

File: src/main/java/smokefree/domain/Initiative.java (hunk 2 above)
Comment (AnandAili) [reply src/main/java/smokefree/domain/Initiative.java]: removed constants from the initiativeProjetion

File: src/main/java/smokefree/domain/Initiative.java (hunk 3 above)
Comment (AnandAili) [reply src/main/java/smokefree/domain/Initiative.java]: DomainExceptions are not directly propagated from AXON to GraphQL DataFetchers. it's translated to ExecutionException That's because these DomainExceptions and ConstraintViolationException are caught in ConfigurableDataFetcherExceptionHandler.java.

File: src/main/java/smokefree/domain/Initiative.java (hunk 4 above)
Comment (AnandAili) [reply src/main/java/smokefree/domain/Initiative.java]: In the earlier discussion, we agreed for 100 Meters. Anyway, it can be changed.

File: src/main/java/smokefree/domain/Initiative.java (hunk 5 above)
Comment (AnandAili) [reply src/main/java/smokefree/domain/Initiative.java]: filter method needs a boolean as a return value:

```.filter( playgroundEntry -> {
                    Playground playground = playgroundEntry.getValue();
                    GlobalPosition currentPlaygroundPosition = new GlobalPosition(playground.getLat() , playground.getLng(), 0.0);
                    double playgroundsDistance = geodeticCalculator.calculateGeodeticCurve(ellipsoidsReference, currentPlaygroundPosition, newPlaygroundPosition).getEllipsoidalDistance();
                    return  playgroundsDistance < distance ? true:false;
                })

File: src/main/java/smokefree/projection/InitiativeProjection.java (hunk 11 above)
Comment (AnandAili) [reply src/main/java/smokefree/projection/InitiativeProjection.java]: yes, removed the code

File: src/main/java/smokefree/projection/InitiativeProjection.java (hunk 13 above)
Comment (AnandAili) [reply src/main/java/smokefree/projection/InitiativeProjection.java]: made changes as suggested.

File: src/main/java/smokefree/domain/Initiative.java (hunk 1 above)
Comment (AnandAili) [reply src/main/java/smokefree/domain/Initiative.java]: unable to find this.

File: src/main/java/smokefree/domain/Initiative.java (hunk 3 above)
Comment (dimitriirtimid) [reply src/main/java/smokefree/domain/Initiative.java]: Ok, but how are these exceptions presented to the frontend?

File: src/main/java/smokefree/domain/Initiative.java (hunk 4 above)
Comment (dimitriirtimid) [reply src/main/java/smokefree/domain/Initiative.java]: Sorry, wrong answer :(
 Have a good look and try again.

File: src/main/resources/bootstrap.yml [hunk 18]
```diff
@@ -101,7 +101,7 @@ jackson:
 aws:
   cognito:
     region: eu-west-1
-    userpoolid: eu-west-1_64ShtC2tS
+    userpoolid: eu-west-1_jnPKAXCKm
```
Comment (dimitriirtimid) [inline src/main/resources/bootstrap.yml]: Why did you change this?

Comment (dimitriirtimid) [reply src/main/resources/bootstrap.yml]: I just checked: I use eu-west-1_WsTxYUHyC when testing from localhost (both on server and frontend of course)

File: src/main/java/smokefree/domain/Initiative.java (hunk 3 above)
Comment (AnandAili) [reply src/main/java/smokefree/domain/Initiative.java]: the exceptions presented as errors,
```result.put("errors", executionResult
                .getErrors()
                .stream()
                .map(error -> determineErrorAttributes(error))
                .collect(toList()));

File: src/main/resources/bootstrap.yml (hunk 18 above)
Comment (AnandAili) [reply src/main/resources/bootstrap.yml]: Earlier I was using the pool ends with eu-west-1~~~~~z4 which was created in dev. I did not see it yesterday, Maybe it is deleted. In order to test yesterday's changes, I had to point to some user poll. But I should not have committed this
